src/compute/cast/primitive_to.rs

```diff
@@ -17,21 +17,35 @@ use super::CastOptions;
 pub fn primitive_to_binary<T: NativeType + lexical_core::ToLexical, O: Offset>(
     from: &PrimitiveArray<T>,
 ) -> BinaryArray<O> {
-    let mut buffer = vec![];
-    let builder = from.iter().fold(
-        MutableBinaryArray::<O>::with_capacity(from.len()),
-        |mut builder, x| {
-            match x {
-                Some(x) => {
-                    lexical_to_bytes_mut(*x, &mut buffer);
-                    builder.push(Some(buffer.as_slice()));
-                }
-                None => builder.push_null(),
-            }
-            builder
-        },
-    );
-    builder.into()
+    let mut values: MutableBuffer<u8> = MutableBuffer::with_capacity(from.len());
+    let mut offsets: MutableBuffer<O> = MutableBuffer::with_capacity(from.len() + 1);
+    offsets.push(O::default());
+
+    let mut offset: usize = 0;
+
+    unsafe {
+        for x in from.values().iter() {
+            values.reserve(offset + T::FORMATTED_SIZE_DECIMAL);
+
+            let bytes = std::slice::from_raw_parts_mut(
+                values.as_mut_ptr().add(offset),
+                values.capacity() - offset,
+            );
+            let len = lexical_core::write_unchecked(*x, bytes).len();
+
+            offset += len;
+            offsets.push(O::from_isize(offset as isize).unwrap());
+        }
+        values.set_len(offset);
+        values.shrink_to_fit();
+    }
+
+    BinaryArray::<O>::from_data(
```

Comment:
```suggestion
    BinaryArray::<O>::from_data_unchecked(
```

since the invariants are uphold above